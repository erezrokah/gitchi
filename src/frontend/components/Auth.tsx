import * as React from 'react';
import { useRef, useState } from 'react';
import cryptoRandomString = require('crypto-random-string');
import { Segment, Button, Message } from 'semantic-ui-react';
import * as storage from '../utils/storage';

const encodeQueryData = (data: Record<string, string>) => {
  const args = [];
  for (const d in data) {
    args.push(encodeURIComponent(d) + '=' + encodeURIComponent(data[d]));
  }
  return args.join('&');
};

const authEndpoint = process.env.REACT_APP_AUTH_ENDPOINT || '';
const origin = authEndpoint.substring(0, authEndpoint.lastIndexOf('/'));

interface Data {
  token: string;
}

interface Event {
  origin: string;
  data: string;
}

interface Props {
  onAuthSuccess: () => void;
}

const ErrorMessage = () => {
  return (
    <Message negative>
      <Message.Header>
        Failed to sign in to GitHub, please try again
      </Message.Header>
    </Message>
  );
};

export const Auth = ({ onAuthSuccess }: Props) => {
  const [error, setError] = useState<Error>();
  const authWindowRef = useRef<Window | null>(null);

  const callback = (error: Error | null, data: Data = { token: '' }) => {
    if (error) {
      console.log(error);
      setError(error);
    } else {
      const { token } = data;
      storage.set({ token }).then(() => onAuthSuccess());
    }
  };

  const authorizeCallback = () => {
    const fn = (e: Event) => {
      if (e.origin !== origin) {
        return;
      }

      if (e.data.indexOf('authorization:github:success:') === 0) {
        const match = e.data.match(
          new RegExp('^authorization:github:success:(.+)$'),
        );
        if (match) {
          const data = JSON.parse(match[1]) as Data;
          window.removeEventListener('message', fn, false);
          authWindowRef.current && authWindowRef.current.close();
          callback(null, data);
        } else {
          callback(new Error("Can't get github success data"));
        }
      }
      if (e.data.indexOf('authorization:github:error:') === 0) {
        const match = e.data.match(
          new RegExp('^authorization:github:error:(.+)$'),
        );
        if (match) {
          const err = JSON.parse(match[1]);
          window.removeEventListener('message', fn, false);
          authWindowRef.current && authWindowRef.current.close();
          callback(new Error(err));
        } else {
          callback(new Error("Can't get github error"));
        }
      }
    };
    return fn;
  };

  const handshakeCallback = () => {
    const fn = (e: Event) => {
      if (e.data === 'authorizing:github' && e.origin === origin) {
        window.removeEventListener('message', fn, false);
        window.addEventListener('message', authorizeCallback(), false);
        return (
          authWindowRef.current &&
          authWindowRef.current.postMessage(e.data, e.origin)
        );
      }
    };
    return fn;
  };

  const openAuthWindow = () => {
    const width = 960;
    const height = 600;
    const left = screen.width / 2 - width / 2;
    const top = screen.height / 2 - height / 2;

    window.addEventListener('message', handshakeCallback(), false);

    const url = `https://github.com/login/oauth/authorize?${encodeQueryData({
      client_id: process.env.REACT_APP_OAUTH_CLIENT_ID || '',
      state: cryptoRandomString({ length: 32, type: 'url-safe' }),
      allow_signup: `${false}`,
      scope: 'public_repo',
    })}`;

    const authWindow = window.open(
      url,
      'Gitchi Authorization',
      `width=${width}, height=${height}, top=${top}, left=${left}`,
    );

    authWindowRef.current = authWindow;
    authWindowRef.current && authWindowRef.current.focus();
  };

  return (
    <Segment basic textAlign="center">
      <Button primary onClick={openAuthWindow}>
        Sign in to Gitchi
      </Button>
      {error ? <ErrorMessage /> : null}
    </Segment>
  );
};
