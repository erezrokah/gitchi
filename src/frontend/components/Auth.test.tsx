/* eslint-disable @typescript-eslint/no-var-requires */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

process.env.REACT_APP_OAUTH_CLIENT_ID = 'REACT_APP_OAUTH_CLIENT_ID';
process.env.REACT_APP_AUTH_ENDPOINT = 'REACT_APP_AUTH_ENDPOINT/';

import { Auth, Event } from './Auth';

jest.mock('crypto-random-string');
jest.mock('../utils/storage', () => {
  return {
    set: jest.fn(() => Promise.resolve()),
  };
});

jest.spyOn(console, 'log').mockImplementation(() => {});

describe('Auth', () => {
  Object.assign(window, { width: 960 * 2 });
  Object.assign(window, { height: 600 * 2 });

  const randomString = 'randomString';
  const cryptoRandomString = require('crypto-random-string') as jest.Mock;
  cryptoRandomString.mockReturnValue(randomString);

  const { set }: { set: jest.Mock } = require('../utils/storage');

  const authWindow = {
    focus: jest.fn(),
    postMessage: jest.fn(),
    close: jest.fn(),
  };

  const addEventListener = jest.fn();
  const removeEventListener = jest.fn();
  window.addEventListener = addEventListener;
  window.removeEventListener = removeEventListener;

  const open = jest.fn();
  window.open = open;

  open.mockReturnValue(authWindow);

  const props = { onAuthSuccess: jest.fn() };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', () => {
    const { asFragment } = render(<Auth {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should register to message event and open auth window on Sign in button click', () => {
    const { getByText } = render(<Auth {...props} />);

    fireEvent.click(getByText('Sign in to Gitchi'));

    expect(window.addEventListener).toHaveBeenCalledWith(
      'message',
      expect.any(Function),
      false,
    );

    expect(window.open).toHaveBeenCalledTimes(1);
    expect(window.open).toHaveBeenCalledWith(
      'https://github.com/login/oauth/authorize?client_id=REACT_APP_OAUTH_CLIENT_ID&state=randomString&allow_signup=false&scope=public_repo',
      'Gitchi Authorization',
      'width=960, height=600, top=-300, left=-480',
    );
    expect(authWindow.focus).toHaveBeenCalledTimes(1);
  });

  test('should handle handshake on "authorizing:github" message with correct origin', () => {
    const { getByText } = render(<Auth {...props} />);

    fireEvent.click(getByText('Sign in to Gitchi'));

    const fn = addEventListener.mock.calls[4][1];
    jest.clearAllMocks();

    const event: Event = {
      origin: 'REACT_APP_AUTH_ENDPOINT',
      data: 'authorizing:github',
    };
    fn(event);

    expect(removeEventListener).toHaveBeenCalledWith('message', fn, false);

    expect(window.addEventListener).toHaveBeenCalledWith(
      'message',
      expect.any(Function),
      false,
    );

    expect(authWindow.postMessage).toHaveBeenCalledTimes(1);
    expect(authWindow.postMessage).toHaveBeenCalledWith(
      event.data,
      event.origin,
    );
  });

  test('should not handle handshake on "authorizing:github" message with wrong origin', () => {
    const { getByText } = render(<Auth {...props} />);

    fireEvent.click(getByText('Sign in to Gitchi'));

    const fn = addEventListener.mock.calls[4][1];
    jest.clearAllMocks();

    const event: Event = {
      origin: 'wrong origin',
      data: 'authorizing:github',
    };
    fn(event);

    expect(removeEventListener).toHaveBeenCalledTimes(0);
    expect(addEventListener).toHaveBeenCalledTimes(0);
    expect(authWindow.postMessage).toHaveBeenCalledTimes(0);
  });

  test('should handle "authorization:github:success" message', async () => {
    const { getByText } = render(<Auth {...props} />);

    fireEvent.click(getByText('Sign in to Gitchi'));

    let fn = addEventListener.mock.calls[4][1];
    jest.clearAllMocks();

    let event: Event = {
      origin: 'REACT_APP_AUTH_ENDPOINT',
      data: 'authorizing:github',
    };
    fn(event);

    const setPromise = Promise.resolve();
    set.mockReturnValue(setPromise);

    const data = { token: 'token' };
    fn = addEventListener.mock.calls[0][1];
    event = {
      origin: 'REACT_APP_AUTH_ENDPOINT',
      data: `authorization:github:success:${JSON.stringify(data)}`,
    };
    fn(event);

    await setPromise;

    expect(authWindow.close).toHaveBeenCalledTimes(1);

    expect(removeEventListener).toHaveBeenCalledWith('message', fn, false);

    expect(set).toHaveBeenCalledTimes(1);
    expect(set).toHaveBeenCalledWith({ token: data.token });
  });

  test('should handle "authorization:github:error" message', async () => {
    const { getByText, asFragment } = render(<Auth {...props} />);
    await act(async () => {
      fireEvent.click(getByText('Sign in to Gitchi'));

      let fn = addEventListener.mock.calls[4][1];
      jest.clearAllMocks();

      let event: Event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: 'authorizing:github',
      };
      fn(event);

      const setPromise = Promise.resolve();
      set.mockReturnValue(setPromise);

      const message = 'error';
      fn = addEventListener.mock.calls[0][1];
      event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: `authorization:github:error:${JSON.stringify(message)}`,
      };
      fn(event);

      await setPromise;

      expect(authWindow.close).toHaveBeenCalledTimes(1);

      expect(removeEventListener).toHaveBeenCalledWith('message', fn, false);

      expect(set).toHaveBeenCalledTimes(0);

      expect(console.log).toHaveBeenCalledTimes(1);
      expect(console.log).toHaveBeenCalledWith(new Error(message));
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test('should not handle authorize on "authorizing:github:success" message with wrong origin', () => {
    const { getByText } = render(<Auth {...props} />);

    fireEvent.click(getByText('Sign in to Gitchi'));

    let fn = addEventListener.mock.calls[4][1];
    jest.clearAllMocks();

    let event: Event = {
      origin: 'REACT_APP_AUTH_ENDPOINT',
      data: 'authorizing:github',
    };
    fn(event);

    const data = { token: 'token' };
    fn = addEventListener.mock.calls[0][1];
    event = {
      origin: 'wrong origin',
      data: `authorization:github:success:${JSON.stringify(data)}`,
    };

    jest.clearAllMocks();

    fn(event);

    expect(removeEventListener).toHaveBeenCalledTimes(0);
    expect(addEventListener).toHaveBeenCalledTimes(0);
    expect(authWindow.postMessage).toHaveBeenCalledTimes(0);
  });

  test('should set error on "authorization:github:success" with no data', () => {
    const { getByText, asFragment } = render(<Auth {...props} />);
    act(() => {
      fireEvent.click(getByText('Sign in to Gitchi'));

      let fn = addEventListener.mock.calls[4][1];
      jest.clearAllMocks();

      let event: Event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: 'authorizing:github',
      };
      fn(event);

      fn = addEventListener.mock.calls[0][1];
      event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: 'authorization:github:success:',
      };
      fn(event);

      expect(console.log).toHaveBeenCalledTimes(1);
      expect(console.log).toHaveBeenCalledWith(
        new Error("Can't get github success data"),
      );
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test('should set error on "authorization:github:error" with no error', () => {
    const { getByText, asFragment } = render(<Auth {...props} />);
    act(() => {
      fireEvent.click(getByText('Sign in to Gitchi'));

      let fn = addEventListener.mock.calls[4][1];
      jest.clearAllMocks();

      let event: Event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: 'authorizing:github',
      };
      fn(event);

      fn = addEventListener.mock.calls[0][1];
      event = {
        origin: 'REACT_APP_AUTH_ENDPOINT',
        data: 'authorization:github:error:',
      };
      fn(event);

      expect(console.log).toHaveBeenCalledTimes(1);
      expect(console.log).toHaveBeenCalledWith(
        new Error("Can't get github error"),
      );
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
