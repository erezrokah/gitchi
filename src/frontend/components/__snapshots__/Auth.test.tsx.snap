// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth should handle "authorization:github:error" message 1`] = `
<DocumentFragment>
  <div
    class="ui basic center aligned segment"
  >
    <button
      class="ui primary button"
    >
      Sign in to Gitchi
    </button>
    <div
      class="ui negative message"
    >
      <div
        class="header"
      >
        Failed to sign in to GitHub, please try again
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Auth should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui basic center aligned segment"
  >
    <button
      class="ui primary button"
    >
      Sign in to Gitchi
    </button>
  </div>
</DocumentFragment>
`;

exports[`Auth should set error on "authorization:github:error" with no error 1`] = `
<DocumentFragment>
  <div
    class="ui basic center aligned segment"
  >
    <button
      class="ui primary button"
    >
      Sign in to Gitchi
    </button>
    <div
      class="ui negative message"
    >
      <div
        class="header"
      >
        Failed to sign in to GitHub, please try again
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Auth should set error on "authorization:github:success" with no data 1`] = `
<DocumentFragment>
  <div
    class="ui basic center aligned segment"
  >
    <button
      class="ui primary button"
    >
      Sign in to Gitchi
    </button>
    <div
      class="ui negative message"
    >
      <div
        class="header"
      >
        Failed to sign in to GitHub, please try again
      </div>
    </div>
  </div>
</DocumentFragment>
`;
